File: MoronBot-Functions/Python/PythonInterface.py
import sys

from MBFunctionInterface import BotMessage
from CwIRC import IRCResponse, ResponseType
from MBUtilities import Settings

import Function

functions = {}

def ProcessMessage(message):
	responses = []
	
	if message.Command == "pyload":
		path = message.ParameterList[0]
		try:
			loadType = LoadFunction(path)
			responses.append( IRCResponse(ResponseType.Say, "Python Function '" + path + "' " + loadType + "oaded!", message.ReplyTo) )
		except Exception, x:
			responses.append( IRCResponse(ResponseType.Say, "Python Load Error: " + str(sys.exc_info()), Settings.Instance.Owner ) )
	elif message.Command == "pyunload":
		path = message.ParameterList[0]
		try:
			success = UnloadFunction(path)
			if success:
				responses.append( IRCResponse(ResponseType.Say, "Python Function '" + path + "' unloaded!", message.ReplyTo) )
			else:
				responses.append( IRCResponse(ResponseType.Say, "Python Function '" + path + "' not found", message.ReplyTo) )
		except Exception, x:
			responses.append( IRCResponse(ResponseType.Say, "Python Unload Error: " + str(sys.exc_info()), Settings.Instance.Owner) )
	else:
		for name, func in functions.iteritems():
			try:
				responses.append( func.GetResponse(message) )
			except Exception, x:
				responses.append( IRCResponse(ResponseType.Say, "Python Execution Error: " + str(sys.exc_info()), Settings.Instance.Owner) )
	
	return responses

def LoadFunction(path, loadAs=""):
	loadType = "l"
	name = path
	src = __import__(name)
	if loadAs != "":
		name = loadAs
	if name in functions:
		loadType = "rel"
	reload(src)
		
	func = src.Instantiate()
		
	functions.update({name:func})
		
	return loadType

def UnloadFunction(name):
	success = 1
	if name in functions.keys():
		del functions[name]
	else:
		success = 0
	
	return success

File: MoronBot-Functions/Python/PythonInterface.cs
using System;
using System.Collections.Generic;
using System.IO;
using System.Linq;
using System.Text;
using System.Text.RegularExpressions;

using CwIRC;
using MBFunctionInterface;

using MBUtilities;

using IronPython.Hosting;
using Microsoft.Scripting.Hosting;
using IronPython.Runtime.Operations;
using Microsoft.Scripting.Runtime;

namespace Utility
{
    public class PythonInterface : Function
    {
        private ScriptEngine engine;
        private ScriptScope scope;
        private ScriptSource source;

        public PythonInterface()
        {
            Help = "A C# function which passes messages to other functions written in python.";
            Type = Types.Regex;
            AccessLevel = AccessLevels.Anyone;

            ReloadScript();
        }

        public override List<IRCResponse> GetResponse(BotMessage message)
        {
            List<IRCResponse> responses = new List<IRCResponse>();

            if (Regex.IsMatch(message.Command, @"pyrestart"))
            {
                ReloadScript();
                responses.Add(new IRCResponse(ResponseType.Say, "PythonInterface reloaded!", message.ReplyTo));
            }
            else
            {
                try
                {
                    IList<object> pyResponses = (IList<object>)engine.Operations.Invoke(scope.GetVariable("ProcessMessage"), message);
                    if (pyResponses.Count > 0)
                        responses = pyResponses.Cast<IRCResponse>().ToList();
                }
                catch (System.Exception ex)
                {
                    Logger.Write("Python Error: " + ex.Message, Settings.Instance.ErrorFile);
                    DynamicStackFrame[] frames = PythonOps.GetDynamicStackFrames(ex);
                    foreach (DynamicStackFrame frame in frames)
                        Logger.Write("\t" +
                            frame.GetFileName() + " " +
                            frame.GetFileLineNumber() + " " +
                            frame.GetMethodName(), Settings.Instance.ErrorFile);
                }
            }

            return responses;
        }

        void ReloadScript()
        {
            engine = Python.CreateEngine();

            engine.Runtime.LoadAssembly(typeof(Function).Assembly);
            engine.Runtime.LoadAssembly(typeof(BotMessage).Assembly);
            engine.Runtime.LoadAssembly(typeof(IRCResponse).Assembly);
            engine.Runtime.LoadAssembly(typeof(Settings).Assembly);

            ICollection<string> paths = engine.GetSearchPaths();
            paths.Add(Settings.Instance.FunctionPath);
            string pathFunctionPython = Path.Combine(Settings.Instance.FunctionPath, "Python");
            paths.Add(pathFunctionPython);
            paths.Add(Path.Combine(pathFunctionPython, "libs"));
            engine.SetSearchPaths(paths);

            scope = engine.CreateScope();
            source = engine.CreateScriptSourceFromFile(Path.Combine(Settings.Instance.FunctionPath, "PythonInterface.py"));

            try
            {
                source.Execute(scope);
            }
            catch (System.Exception ex)
            {
                Logger.Write("Python Error: " + ex.Message, Settings.Instance.ErrorFile);
                DynamicStackFrame[] frames = PythonOps.GetDynamicStackFrames(ex);
                foreach (DynamicStackFrame frame in frames)
                    Logger.Write("\t" +
                        frame.GetFileName() + " " +
                        frame.GetFileLineNumber() + " " +
                        frame.GetMethodName(), Settings.Instance.ErrorFile);
            }
        }
    }
}
